### PR TITLE
Add dhall to toolchain

### PR DESCRIPTION
**Pull request by balsoft**
Adds dhall to Mina toolchain. Needed for https://github.com/MinaProtocol/mina/pull/10939

### CODE REVIEW

**Review by yorickvP — APPROVED** [review]

**Comment by yorickvP** [comment]
CI Failure is genuine: E: Unable to locate package dhall

File: dockerfiles/stages/3-toolchain [hunk 1]
```diff
@@ -105,3 +104,17 @@ RUN curl -sL https://deb.nodesource.com/setup_14.x | sudo -E bash - \
     && echo "deb https://dl.yarnpkg.com/debian/ stable main" | sudo tee /etc/apt/sources.list.d/yarn.list \
     && sudo apt update \
     && sudo apt install -y nodejs yarn
+
+# Dhall
+RUN curl -sL https://github.com/dhall-lang/dhall-haskell/releases/download/1.41.1/dhall-1.41.1-x86_64-linux.tar.bz2 > /tmp/dhall.tar.bz2
+    && mkdir -p /tmp/dhall
+    && tar -jz -C /tmp/dhall -f /tmp/dhall.tar.bz2
+    && sudo mv /tmp/dhall/bin/dhall /usr/bin
```
Comment (yorickvP) [inline dockerfiles/stages/3-toolchain:112]:
```suggestion
    && sudo tar --extract --file=/tmp/dhall.tar.bz2 --bzip2 --directory=/usr bin/dhall
```

File: dockerfiles/stages/3-toolchain [hunk 2]
```diff
@@ -105,3 +104,17 @@ RUN curl -sL https://deb.nodesource.com/setup_14.x | sudo -E bash - \
...
+    && curl -sL https://github.com/dhall-lang/dhall-haskell/releases/download/1.41.1/dhall-bash-1.0.40-x86_64-linux.tar.bz2 > /tmp/dhall-bash.tar.bz2
+    && mkdir -p /tmp/dhall-bash
+    && tar -jz -C /tmp/dhall-bash -f /tmp/dhall-bash.tar.bz2
+    && sudo mv /tmp/dhall-bash/bin/dhall-to-bash /usr/bin
```
Comment (yorickvP) [inline dockerfiles/stages/3-toolchain:116]:
```suggestion
    && sudo tar --extract --file=/tmp/dhall-bash.tar.bz2 --bzip2 --directory=/usr bin/dhall-to-bash
```

File: dockerfiles/stages/3-toolchain [hunk 3]
```diff
@@ -105,3 +104,17 @@ RUN curl -sL https://deb.nodesource.com/setup_14.x | sudo -E bash - \
...
+    && curl -sL https://github.com/dhall-lang/dhall-haskell/releases/download/1.41.1/dhall-yaml-1.2.10-x86_64-linux.tar.bz2 > /tmp/dhall-yaml.tar.bz2
+    && mkdir -p /tmp/dhall-yaml
+    && tar -jz -C /tmp/dhall-yaml -f /tmp/dhall-yaml.tar.bz2
+    && sudo mv /tmp/dhall-yaml/bin/dhall-to-yaml /usr/bin
```
Comment (yorickvP) [inline dockerfiles/stages/3-toolchain:120]:
```suggestion
    && sudo tar --extract --file=/tmp/dhall-yaml.tar.bz2 --bzip2 --directory=/usr bin/dhall-to-yaml
```

**Comment by yorickvP** [comment]
> Error response from daemon: Dockerfile parse error line 305: unknown instruction: &&

Need backslashes at the end of the lines

**Comment by yorickvP** [comment]
lgtm

File: dockerfiles/stages/3-toolchain [hunk 4]
```diff
@@ -104,3 +104,11 @@ RUN curl -sL https://deb.nodesource.com/setup_14.x | sudo -E bash - \
     && echo "deb https://dl.yarnpkg.com/debian/ stable main" | sudo tee /etc/apt/sources.list.d/yarn.list \
     && sudo apt update \
     && sudo apt install -y nodejs yarn
+
+# Dhall
+RUN curl -sL https://github.com/dhall-lang/dhall-haskell/releases/download/1.41.1/dhall-1.41.1-x86_64-linux.tar.bz2 > /tmp/dhall.tar.bz2 \
```
[the remark this answers is not in the document]
Comment (lk86) [reply dockerfiles/stages/3-toolchain:109]: Also can you add a new DHALL_VERSION=1.41.1 to the top of this file and potentially the DHALL_JSON_VERSION and DHALL_BASH_VERSION

**Review by lk86 — CHANGES_REQUESTED** [review]

File: dockerfiles/stages/3-toolchain (hunk 4 above)
[the remark this answers is not in the document]
Comment (yorickvP) [reply dockerfiles/stages/3-toolchain:109]: We could do `curl | tar` to avoid the temporary file

**Review by lk86 — APPROVED** [review]

**Pull request merged**